File: stage3/s3_pipe_06.js
// S3_PIPE_06 — Safe pipeline
function pipeSafe(...fns) {
  return function (value) {
    try {
      const result = fns.reduce((acc, fn) => fn(acc), value);
      return { ok: true, value: result };
    } catch (error) {
      return { ok: false, error };
    }
  };
}

// Tests
const parseNumber = (x) => {
  const n = Number(x);
  if (Number.isNaN(n)) throw new Error("Not a number");
  return n;
};

const double = (x) => x * 2;

const safeProcess = pipeSafe(parseNumber, double);

console.log(safeProcess("10"));
// { ok: true, value: 20 }

console.log(safeProcess("abc"));
// { ok: false, error: Error: Not a number }

File: stage3/C/S3_PIPE_06.py
# S3_PIPE_06.py

def pipe_safe(*fns):
    def run(x):
        try:
            for fn in fns:
                x = fn(x)
            return {"ok": True, "value": x}
        except Exception as e:
            return {"ok": False, "error": str(e)}
    return run


def main():
    def must_be_int(x):
        if not isinstance(x, int):
            raise ValueError("expected int")
        return x

    safe = pipe_safe(
        must_be_int,
        lambda n: n * 2,
        lambda n: n + 1,
    )

    print(safe(10))
    print(safe("10"))


if __name__ == "__main__":
    main()
